tests/frontend/test_parameters.py

```diff
@@ -11,118 +11,156 @@
+
+SCALAR_TEST_VALUES = [3, 0.14, 4.2 + 0.5j]
+TEST_VALUES = SCALAR_TEST_VALUES + [np.array([0.1, 0.987654])]
+
+
+def binary_arithmetic(pp, qq, p, q):
+    """Test the correctness of basic binary arithmetic expressions."""
+    assert par_evaluate(pp + qq) == pytest.approx(p + q)
+    assert par_evaluate(pp - qq) == pytest.approx(p - q)
+    assert par_evaluate(pp * qq) == pytest.approx(p * q)
+    assert par_evaluate(pp / qq) == pytest.approx(p / q)
+    assert par_evaluate(pp ** qq) == pytest.approx(p ** q)
+
+
+class TestParameter:
+    """Basic parameter functionality."""
+
+    def test_measured_par(self):
+        """Compare two ways of creating measured parameter instances from RegRefs."""
+        r = RegRef(0)
+        s = RegRef(1)
+        p = MeasuredParameter(r)
+        assert p == r.par
+        assert p != s.par
+
+    @pytest.mark.parametrize("r", TEST_VALUES)
+    def test_par_is_symbolic(self, r):
+        """Recognizing symbolic parameters."""
+        p = FreeParameter('x')
+        q = MeasuredParameter(RegRef(0))
+
+        assert not par_is_symbolic(r)
+        assert par_is_symbolic(q)
+        assert par_is_symbolic(p)
+        assert par_is_symbolic(pf.sin(p))
+        assert par_is_symbolic(p + r)
+        assert par_is_symbolic(p - r)
+        assert par_is_symbolic(p * r)
+        assert par_is_symbolic(p / r)
+        assert par_is_symbolic(p ** r)
+        assert par_is_symbolic(p - p)  # no simplification
+
+        # object array with symbols
+        a = np.array([[0.1, 3, 0], [0.3, 2, p], [1, 2, 4]])
+        assert a.dtype == object
+        assert par_is_symbolic(a)
+
+        # object array, no symbols
+        a = np.array([[0.1, 3, 0], [0.3, 2, 0], [1, 2, 4]], dtype=object)
+        assert a.dtype == object
+        assert not par_is_symbolic(a)
+
+        # float array, no symbols
+        a = np.array([[0.1, 3, 0], [0.3, 2, 0], [1, 2, 4]])
+        assert a.dtype != object
+        assert not par_is_symbolic(a)
+
+
+    def test_par_regref_deps(self):
+        """RegRef dependencies of parameters."""
+        r = [RegRef(k) for k in range(2)]
+        R0 = set([r[0]])
+        R = set(r)
+
+        s = 0.4
+        x = FreeParameter('x')
+        p = MeasuredParameter(r[0])
+        q = MeasuredParameter(r[1])
+
+        assert par_regref_deps(s) == set()
+        assert par_regref_deps(x) == set()
+        assert par_regref_deps(p) == R0
+        assert par_regref_deps(s * p) == R0
+        assert par_regref_deps(x * p) == R0
+        assert par_regref_deps(p + q) == R
+        assert par_regref_deps(s * p + q ** x) == R
+
+    def test_par_str(self):
+        """String representations of parameters."""
+
+        a = 0.1234567
+        b = np.array([0, 0.987654])
+        c = MeasuredParameter(RegRef(1))
+        d = FreeParameter('x')
+
+        assert par_str(a) == '0.1235'  # rounded to 4 decimals
+        assert par_str(b) == '[0.     0.9877]'
+        assert par_str(c) == 'q[1].par'
```

Comment:
At the same time, I don't think the users variable name will ever be accessible. I think this is acceptable, especially because we use `q` in almost all documentation.

Note: In 'pure' blackbird, we force printed regref names to be q0, q1, etc. Might be worth doing the same in SF?

Reply:
Might make sense to not make it look like a python object. Just stick to `q0`,`q1`, etc

Reply:
This is a slightly larger issue, `RegRef.__str__` also returns 'q[1]` etc at the moment.

Reply:
I changed it, and left Regref.__str__ as it was.